src/mqtt.rs

```diff
@@ -13,14 +15,221 @@ pub struct Mqtt {
     client: Client,
 }
 
+
+/// `HMSStateResponse` is a struct that contains the data from the inverter.
+///
+/// Provide utility functions to extract data from the struct.
+impl HMSStateResponse {
+    fn get_model(&self) -> String {
+         // TODO: identify model from dtu_sn
+        "HMS-800W-T2".to_string()
+    }
+
+    fn get_name(&self) -> String {
+        format!("Hoymiles {} {}", self.get_model(), self.short_dtu_sn())
+    }
+
+    fn short_dtu_sn(&self) -> String {
+        self.dtu_sn.iter().take(8).map(|v| v.to_string()).collect::<String>()
+    }
+
+    fn get_total_efficiency(&self) -> f32{
+        let total_input_power: f32 = self.port_state.iter()
+        .map(|port| port.pv_power as f32).sum();
+    
+        if total_input_power > 0.0 {
+            return (self.pv_current_power as f32) / total_input_power * 100.0
+        } else {
+            return 0.0
+        };
+    }
+
+    fn get_inverter(&self, pv_port: i32) -> InverterState {
+        self.inverter_state.iter().find(
+            |inv| inv.port_id == pv_port
+        ).unwrap_or(&InverterState::default()).clone()
+    }
+
+    fn to_json_payload(&self) -> serde_json::Value {
```

Comment:
consider using the [protobuf-json](https://crates.io/crates/protobuf-json) crate, which makes this function a one-liner basically.